File: basketball-csharp/BasketballTickets/Service/ServiceImpl.cs
using BasketballTickets.Domain;
using BasketballTickets.Repository.Interfaces;
using log4net;

namespace BasketballTickets.Service;

/// <summary>
/// Implementeaza toata logica de business pentru sistemul de bilete de baschet.
/// Controllerele primesc o referinta la aceasta clasa — nu ating niciodata repository-urile direct.
/// Echivalent cu ServiceImpl.java
/// </summary>
public class ServiceImpl : IService
{
    private static readonly ILog Logger = LogManager.GetLogger(typeof(ServiceImpl));

    private readonly ICashierRepository _cashierRepo;
    private readonly IMatchRepository _matchRepo;
    private readonly ICustomerRepository _customerRepo;
    private readonly ITicketRepository _ticketRepo;

    // Lista de observatori notificati cand datele unui meci se schimba
    // (pentru update-uri multi-client in timp real)
    private readonly List<IMatchUpdateObserver> _observers = [];

    public ServiceImpl(ICashierRepository cashierRepo,
                       IMatchRepository matchRepo,
                       ICustomerRepository customerRepo,
                       ITicketRepository ticketRepo)
    {
        _cashierRepo  = cashierRepo;
        _matchRepo    = matchRepo;
        _customerRepo = customerRepo;
        _ticketRepo   = ticketRepo;
        Logger.Info("ServiceImpl created");
    }

    // ---- Suport Observer (pentru networking/multi-client in viitor) ----

    public void AddObserver(IMatchUpdateObserver observer)    => _observers.Add(observer);
    public void RemoveObserver(IMatchUpdateObserver observer) => _observers.Remove(observer);

    private void NotifyMatchUpdated(Match match) =>
        _observers.ForEach(o => o.OnMatchUpdated(match));

    // ---- Autentificare ----

    public Cashier? Login(string username, string password)
    {
        Logger.Debug($"Login attempt: username={username}");
        return _cashierRepo.FindByUsernameAndPassword(username, password);
    }

    // ---- Meciuri ----

    public IList<Match> GetAllMatches()
    {
        Logger.Debug("GetAllMatches()");
        return _matchRepo.FindAll();
    }

    // ---- Vanzare bilete ----

    public Ticket SellTicket(string customerName, string customerAddress,
                             long matchId, int numberOfSeats)
    {
        Logger.Debug($"SellTicket: customer='{customerName}', matchId={matchId}, seats={numberOfSeats}");

        // 1. Valideaza meciul si locurile disponibile
        var match = _matchRepo.FindById(matchId)
            ?? throw new InvalidOperationException($"Match not found: {matchId}");

        if (match.AvailableSeats < numberOfSeats)
        {
            Logger.Warn($"Not enough seats for matchId={matchId}: " +
                        $"requested={numberOfSeats}, available={match.AvailableSeats}");
            throw new InvalidOperationException(
                $"Not enough seats available. Requested: {numberOfSeats}, " +
                $"Available: {match.AvailableSeats}");
        }

        // 2. Gaseste sau creeaza clientul
        var existing = _customerRepo.FindByNameOrAddress(customerName, customerAddress);
        Customer customer;
        if (existing.Count == 0)
        {
            customer = _customerRepo.Save(new Customer(0, customerName, customerAddress));
            Logger.Info($"New customer created: {customer}");
        }
        else
        {
            customer = existing[0];
            Logger.Debug($"Existing customer found: {customer}");
        }

        // 3. Salveaza biletul
        var ticket = _ticketRepo.Save(new Ticket(0, customer, match, numberOfSeats));

        // 4. Decrementeaza locurile disponibile
        var updated = _matchRepo.UpdateAvailableSeats(matchId, -numberOfSeats);

        // 5. Notifica toti clientii conectati (Observer pattern)
        NotifyMatchUpdated(updated);

        Logger.Info($"Ticket sold: {ticket}");
        return ticket;
    }

    // ---- Cautare ----

    public IList<Ticket> SearchTickets(string? customerName, string? customerAddress)
    {
        Logger.Debug($"SearchTickets(name='{customerName}', address='{customerAddress}')");
        return _ticketRepo.FindByCustomerNameOrAddress(customerName, customerAddress);
    }

    // ---- Modificare ----

    public Ticket ModifyTicketSeats(long ticketId, int newNumberOfSeats)
    {
        Logger.Debug($"ModifyTicketSeats(ticketId={ticketId}, newSeats={newNumberOfSeats})");

        var ticket = _ticketRepo.FindById(ticketId)
            ?? throw new InvalidOperationException($"Ticket not found: {ticketId}");

        var oldSeats = ticket.NumberOfSeats;
        var delta    = oldSeats - newNumberOfSeats; // pozitiv = eliberam locuri, negativ = avem nevoie de mai multe

        // Daca avem nevoie de mai multe locuri, validam disponibilitatea
        if (delta < 0)
        {
            var match = _matchRepo.FindById(ticket.Match.Id)
                ?? throw new InvalidOperationException($"Match not found: {ticket.Match.Id}");

            if (match.AvailableSeats < -delta)
            {
                Logger.Warn($"Cannot increase seats for ticketId={ticketId}: " +
                            $"need {-delta} more, only {match.AvailableSeats} available");
                throw new InvalidOperationException(
                    $"Not enough seats available. Need {-delta} more, " +
                    $"but only {match.AvailableSeats} available.");
            }
        }

        ticket.NumberOfSeats = newNumberOfSeats;
        _ticketRepo.Update(ticket);

        var updated = _matchRepo.UpdateAvailableSeats(ticket.Match.Id, delta);
        NotifyMatchUpdated(updated);

        Logger.Info($"Ticket {ticketId} modified: {oldSeats} → {newNumberOfSeats} seats");
        return ticket;
    }
}

File: java/src/main/java/org/example/service/ServiceImpl.java
package org.example.service;


import org.example.domain.Cashier;
import org.example.domain.Customer;
import org.example.domain.Match;
import org.example.domain.Ticket;
import org.example.repository.interfaces.CashierRepository;
import org.example.repository.interfaces.CustomerRepository;
import org.example.repository.interfaces.MatchRepository;
import org.example.repository.interfaces.TicketRepository;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.List;
import java.util.Optional;

/**
 * Implements all business logic for the Basketball Ticket System.
 *
 * Controllers receive a reference to this class only — they never
 * talk directly to repositories. This makes it easy to swap the
 * persistence layer or add networking later.
 */
public class ServiceImpl implements Service {

    private static final Logger logger = LogManager.getLogger(ServiceImpl.class);

    private final CashierRepository cashierRepo;
    private final MatchRepository matchRepo;
    private final CustomerRepository customerRepo;
    private final TicketRepository ticketRepo;

    // List of observers notified when match data changes (for multi-client real-time updates)
    private final List<MatchUpdateObserver> observers = new java.util.concurrent.CopyOnWriteArrayList<>();

    public ServiceImpl(CashierRepository cashierRepo,
                       MatchRepository matchRepo,
                       CustomerRepository customerRepo,
                       TicketRepository ticketRepo) {
        this.cashierRepo = cashierRepo;
        this.matchRepo   = matchRepo;
        this.customerRepo = customerRepo;
        this.ticketRepo  = ticketRepo;
        logger.info("ServiceImpl created");
    }

    // ---- Observer support (for future network/multi-client scale-up) ----

    public void addObserver(MatchUpdateObserver observer) {
        observers.add(observer);
    }

    public void removeObserver(MatchUpdateObserver observer) {
        observers.remove(observer);
    }

    private void notifyMatchUpdated(Match match) {
        observers.forEach(o -> o.onMatchUpdated(match));
    }

    // ---- Auth ----

    @Override
    public Optional<Cashier> login(String username, String password) {
        logger.debug("login attempt: username={}", username);
        return cashierRepo.findByUsernameAndPassword(username, password);
    }

    // ---- Matches ----

    @Override
    public List<Match> getAllMatches() {
        logger.debug("getAllMatches()");
        return matchRepo.findAll();
    }

    // ---- Ticket sales ----

    @Override
    public Ticket sellTicket(String customerName, String customerAddress,
                             Long matchId, int numberOfSeats) {
        logger.debug("sellTicket: customer='{}', matchId={}, seats={}", customerName, matchId, numberOfSeats);

        // 1. Validate match exists and has enough seats
        Match match = matchRepo.findById(matchId)
                .orElseThrow(() -> new IllegalArgumentException("Match not found: " + matchId));

        if (match.getAvailableSeats() < numberOfSeats) {
            logger.warn("Not enough seats for matchId={}: requested={}, available={}",
                    matchId, numberOfSeats, match.getAvailableSeats());
            throw new IllegalArgumentException(
                    "Not enough seats available. Requested: " + numberOfSeats
                            + ", Available: " + match.getAvailableSeats());
        }

        // 2. Find or create customer
        List<Customer> existing = customerRepo.findByNameOrAddress(customerName, customerAddress);
        Customer customer;
        if (existing.isEmpty()) {
            customer = customerRepo.save(new Customer(null, customerName, customerAddress));
            logger.info("New customer created: {}", customer);
        } else {
            customer = existing.get(0);
            logger.debug("Existing customer found: {}", customer);
        }

        // 3. Save ticket
        Ticket ticket = ticketRepo.save(new Ticket(null, customer, match, numberOfSeats));

        // 4. Decrement available seats
        Match updated = matchRepo.updateAvailableSeats(matchId, -numberOfSeats);

        // 5. Notify all connected clients (Observer pattern — ready for networking)
        notifyMatchUpdated(updated);

        logger.info("Ticket sold: {}", ticket);
        return ticket;
    }

    // ---- Search ----

    @Override
    public List<Ticket> searchTickets(String customerName, String customerAddress) {
        logger.debug("searchTickets(name='{}', address='{}')", customerName, customerAddress);
        return ticketRepo.findByCustomerNameOrAddress(customerName, customerAddress);
    }

    // ---- Modify ----

    @Override
    public Ticket modifyTicketSeats(Long ticketId, int newNumberOfSeats) {
        logger.debug("modifyTicketSeats(ticketId={}, newSeats={})", ticketId, newNumberOfSeats);

        Ticket ticket = ticketRepo.findById(ticketId)
                .orElseThrow(() -> new IllegalArgumentException("Ticket not found: " + ticketId));

        int oldSeats = ticket.getNumberOfSeats();
        int delta    = oldSeats - newNumberOfSeats; // positive = freeing seats, negative = needing more

        // If we need more seats, validate availability
        if (delta < 0) {
            Match match = matchRepo.findById(ticket.getMatch().getId()).orElseThrow();
            if (match.getAvailableSeats() < -delta) {
                logger.warn("Cannot increase seats for ticketId={}: need {} more, only {} available",
                        ticketId, -delta, match.getAvailableSeats());
                throw new IllegalArgumentException(
                        "Not enough seats available. Need " + (-delta)
                                + " more, but only " + match.getAvailableSeats() + " available.");
            }
        }

        ticket.setNumberOfSeats(newNumberOfSeats);
        ticketRepo.update(ticket);

        Match updated = matchRepo.updateAvailableSeats(ticket.getMatch().getId(), delta);
        notifyMatchUpdated(updated);

        logger.info("Ticket {} modified: {} → {} seats", ticketId, oldSeats, newNumberOfSeats);
        return ticket;
    }
}
